### ISSUE TITLE
Add --semver option to adhere to package.json ranges

### ISSUE THREAD

**Issue by raineorshine** [raineorshine/npm-check-updates#1054]
Following https://github.com/raineorshine/npm-check-updates/pull/1052#issuecomment-1054125274:

> One last thing - should we somehow add the two filter + target as a new param/cli option? As in `semver = true` sets both - so that people don't need to come up with the two functions I posted in the top?
> 
> ```js
> target: (name, [{ operator }]) => operator === '^' ? 'minor' : operator === '~' ? 'patch' : 'latest',
> filter: (_, [{ major, operator }]) => !(major === '0' || major === undefined || operator === undefined)
> ```

@Primajin Yes, I think that's a great idea.

> ```js
> target: (name, [{ operator }]) => operator === '^' ? 'minor' : operator === '~' ? 'patch' : 'latest',
> ```

There are more cases that need to be handled: Basically we need to support [Ranges](https://github.com/npm/node-semver#ranges), [Advanced Ranges](https://github.com/npm/node-semver#advanced-range-syntax), and [Prerelease Tags](https://github.com/npm/node-semver#prerelease-tags) (which should set the target to `greatest`).

We may want to use [semver.satisfies](https://github.com/npm/node-semver#usage) so that we don't have to implement all of those conditions ourselves.

> filter: (_, [{ major, operator }]) => !(major === '0' || major === undefined || operator === undefined)

We still want to upgrade dependencies like `^0.1.1` according to semver, and handle that in `target`. So wouldn't we just want `filter: (_, [{ operator }]) => operator`?

**Comment by Primajin** [comment]
Yeah, with `0` I'm always unsure since it often has breaking changes in minor versions

**Comment by raineorshine** [comment]
Users can of course specify their own target function, but if we add a `--semver` option it should follow semver exactly. Thus, `^0.1.1` would be upgraded to greater minor versions, even if they are breaking. The user can remove the `^` if they prefer a fixed dependency.

**Comment by Primajin** [comment]
Sure, fine with me 👍🏻

**Comment by benface** [comment]
> Users can of course specify their own target function, but if we add a `--semver` option it should follow semver exactly. Thus, `^0.1.1` would be upgraded to greater minor versions, even if they are breaking. The user can remove the `^` if they prefer a fixed dependency.

Wait, what? In semver, `^0.1.1` would upgrade to 0.1.2, but not to 0.2.0, so I don't understand what you mean?

**Comment by raineorshine** [comment]
You are correct! My bad.

**Comment by raineorshine** [comment]
Initial support for `--target semver` added and published in `v16.11.0`. It will upgrade if there are newer versions without breaking the specified semver version range.

Please create a new issue if it does not work as expected. I'm sure there are some issues with its handling of prerelease versions, so I could use some feedback to get a better sense of the expected behavior.

**Comment by Primajin** [comment]
It's lovely ❤️ thanks!

**Comment by rifandani** [comment]
I have `"axios": "^1.5.0"` as `dependencies` in `package.json`.
When i try to upgrade it using these command `ncu -u --install always --filter axios --target semver`, it correctly updates the `package.json` to `"axios": "^1.5.1"`.
But when i see the actual installed modules in `node_modules`, it showed as `"version": "1.6.2"`

<img width="1176" alt="image" src="https://github.com/raineorshine/npm-check-updates/assets/57568218/34522c42-1d7c-498a-86f4-f72ff4eeccbb">

this is what changes in `pnpm.lock.yaml`

<img width="953" alt="image" src="https://github.com/raineorshine/npm-check-updates/assets/57568218/5c8c4bf5-4b3a-4c12-9c4e-561c18ed6b90">

Is this the correct behavior?

**Comment by ljharb** [comment]
Yes, 1.6.2 satisfies ^1.5.1.

**Comment by Primajin** [comment]
It's just a little odd that it didn't update your package directly to the latest available version still falling under semver

**Comment by rifandani** [comment]
yeah, if the actual installed module is `^1.6.2`, then why in `package.json` does not also updates to `^1.6.2` instead of `^1.5.1` ?

**Comment by raineorshine** [comment]
@rifandani This appears to be a regression with the pnpm integration. Please open a new issue and I will investigate.

```sh
$ ncu --packageManager pnpm --target semver
Checking /Users/raine/projects/ncu-issues/1054/package.json
[====================] 1/1 100%

 axios  ^1.5.0  →  ^1.5.1

Run ncu --packageManager pnpm --target semver -u to upgrade package.json

$ ncu --packageManager npm --target semver
Checking /Users/raine/projects/ncu-issues/1054/package.json
[====================] 1/1 100%

 axios  ^1.5.0  →  ^1.6.2

Run ncu --packageManager npm --target semver -u to upgrade package.json
```

**Comment by raineorshine** [comment]
It was a very dumb mistake. I published a patch to `v16.14.12`. Let me know if that fixes it.

> it correctly updates the `package.json` to `"axios": "^1.5.1"`.

Actually, semver will upgrade to `^1.6.2` because the `^` range operator includes minor versions (exception for major-version zero).

**Comment by rifandani** [comment]
> It was a very dumb mistake. I published a patch to `v16.14.12`. Let me know if that fixes it.

thanks a lot, now package.json also updates correctly to `^1.6.2`

> Actually, semver should upgrade to `^1.6.2` because the `^` range operator includes minor versions (exception for major-version zero).

yeah i was wrong, it should upgrade to `^1.6.2` instead of `^1.5.1`

**Comment by XhmikosR** [comment]
Isn't this issue fixed a long time ago?
